lib/spack/spack/version.py

```diff
@@ -1226,14 +1225,13 @@ def fetcher(self):
     @property
     def repository_uri(self):
         """Identifier for git repos used within the repo and metadata caches."""
-        try:
-            components = [
-                str(c).lstrip("/") for c in spack.util.url.parse_git_url(self.pkg.git) if c
-            ]
-            return os.path.join(*components)
-        except ValueError:
-            # If it's not a git url, it's a local path
-            return os.path.abspath(self.pkg.git)
+        (scheme, netloc, path, params, query, fragment) = urllib.parse.urlparse(self.pkg.git)
+        path = path.lstrip("/")
+        if scheme == "file":
+            return Path("file") / Path(path).name
```

Comment:
I'm not clear why `Path(...).name` is used, since that removes all the path components except for the last:

```
>>> import pathlib
>>> pathlib.Path('x/y/z').name
'z'
```

That seems like different behavior than before (and also wrong)

Reply:
The previous implementation returned the full path to the local git repo. It is later used to concatenate onto a path. This works for linux because the absolute paths start with `/`. This would result in something later of `/my_new_location/file/local/git/repo`. Because windows uses different volume mounts, you end up getting something along the lines of `C:\\my_new_location\\file\\C:local\\git\\repo`. If we simply strip the file mount and use the rest of the path to match linux, we run into windows path length issues when calling git commands on the new path. I had spend a good amount of time on this section and hadn't found a better solution than using `.name` to remove the file mount and enough of the path to make the git commands work. It change functionality from before though. @scheibelp Do you have any thoughts on how to resolve this problem?

Reply:
> It is later used to concatenate onto a path. This works for linux because the absolute paths start with /

and also I assume because we have `lstrip(/)` (which works for Linux but not Windows)

> It is later used to concatenate onto a path. 

(this logic has since moved and is one of the reasons for conflicts, but I think the principle is the same in the new location)

I think by "later" you mean in `lookup_ref`.

If we ignore everything but the final component, we run the risk of caching two repositories with the same identifier. (e.g. if two different users fork a repo). I think it would be safer to hash (i.e. `spack.util.hash.b32_hash`) the full pathname and take the first several characters (to fall within the path length limits on Windows).

Reply:
Notes: 

7 char of hash on the path that would be constructed. Use hash as folder name for later clone. Make windows only for hashing.

Reply:
I added the hashing for windows only and on linux, it will use the full path